yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts

```diff
@@ -104,7 +133,30 @@ export class NoteDataProvider implements DataProvider {
     });
   }
 
-  public removeNotesAfter(blockNumber: number): Promise<void> {
+  /**
+   * Synchronizes notes and nullifiers to a specific block number.
+   *
+   * This method ensures that the state of notes and nullifiers is consistent with the
+   * specified block number. It restores any notes that were nullified after the given block
+   * and deletes any active notes created after that block.
+   *
+   * @param blockNumber - The block number to synchronize to (the reorg point)
+   * @param synchedBlockNumber - Optional upper bound for the block range to process (current tip)
+   */
+  public async syncNotesAndNullifiers(blockNumber: number, synchedBlockNumber?: number): Promise<void> {
```

Comment:
```suggestion
  public async rollbackNotesAndNullifiers(blockNumber: number, synchedBlockNumber: number): Promise<void> {
```

I think the `syncNotesAndNullifiers` name insufficiently communicates that we are dealing with reorgs - here it would make me think we are syncing new notes and nullifiers.

I also checked and the `synchedBlockNumber` is always defined so please update that everywhere (It's a tech debt from time when we didn't throw in `syncDataProvider.getBlockNumber()` function).

Reply:
API-wise I find it quite strange that we have the `synchedBlockNumber` on the input. It feels like a complexity leak out of this class. It also makes me worried that it might make this class more brittle - what if there is a bug somewhere and for whatever reason we store here a note or nullifier beyond the synched block number?.

Maybe we could think of how to redesign the stores here to not need this on the input. Or maybe to have it on the input but use it only to be able to check that there are no notes and nullifers beyond that block number.

This might seem too paranoid but if there was a bug here it would be a very unpleasant thing to debug.